url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/io/DiskStorageManager.java

```diff
@@ -0,0 +1,423 @@
+				}
+
+				// load index table in memory.
+				count = m_indexFile.readInt();
+
+				for (int cCount = 0; cCount < count; cCount++) {
+					Entry e = new Entry();
+
+					id = m_indexFile.readInt();
+					e.m_length = m_indexFile.readInt();
+
+					int count2 = m_indexFile.readInt();
+
+					for (int cCount2 = 0; cCount2 < count2; cCount2++) {
+						page = m_indexFile.readInt();
+						e.m_pages.add(new Integer(page));
+					}
+					m_pageIndex.put(new Integer(id), e);
+				}
+			} catch (EOFException ex) {
+				throw new IllegalStateException("Corrupted index file.");
+			}
+		}
+	}
+
+	/**
+	 * save the index to file
+	 */
+	@Override
+	public void flush() {
+		try {
+			m_indexFile.seek(0l);
+
+			m_indexFile.writeInt(m_pageSize);
+			m_indexFile.writeInt(m_nextPage);
+
+			int id, page;
+			int count = m_emptyPages.size();
+
+			m_indexFile.writeInt(count);
+
+			Iterator<Integer> it = m_emptyPages.iterator();
+			while (it.hasNext()) {
+				page = it.next().intValue();
+				m_indexFile.writeInt(page);
+			}
+
+			count = m_pageIndex.size();
+			m_indexFile.writeInt(count);
+
+			Iterator<Map.Entry<Integer, DiskStorageManager.Entry>> it2 = m_pageIndex
+					.entrySet().iterator();
+
+			while (it2.hasNext()) {
+				Map.Entry<Integer, DiskStorageManager.Entry> me = it2.next();
+				id = me.getKey().intValue();
+				m_indexFile.writeInt(id);
+
+				Entry e = me.getValue();
+				count = e.m_length;
+				m_indexFile.writeInt(count);
+
+				count = e.m_pages.size();
+				m_indexFile.writeInt(count);
+
+				for (int cIndex = 0; cIndex < count; cIndex++) {
+					page = e.m_pages.get(cIndex).intValue();
+					m_indexFile.writeInt(page);
+				}
+			}
+		} catch (IOException ex) {
+			throw new IllegalStateException("Corrupted index file.");
+		}
+	}
+
+	/**
+	 * load all data of the id-th entry (data in e.m_pages)
+	 */
+	@Override
+	public byte[] loadByteArray(final int id) {
+		Entry e = m_pageIndex.get(new Integer(id));
+		if (e == null) {
+			throw new InvalidPageException(id);
+		}
+
+		int cNext = 0;
+		int cTotal = e.m_pages.size();
+
+		byte[] data = new byte[e.m_length];
+		int cIndex = 0;
+		int cLen;
+		int cRem = e.m_length;
+
+		do {
+			try {
+				//locate every page that used by every entry
+				m_dataFile.seek(e.m_pages.get(cNext).intValue() * m_pageSize);
+				int bytesread = m_dataFile.read(m_buffer);
+				if (bytesread != m_pageSize) {
+					throw new IllegalStateException("Corrupted data file.");
+				}
+			} catch (IOException ex) {
+				throw new IllegalStateException("Corrupted data file.");
+			}
+
+			cLen = cRem > m_pageSize ? m_pageSize : cRem;
+			System.arraycopy(m_buffer, 0, data, cIndex, cLen);
+
+			cIndex += cLen;
+			cRem -= cLen;
+			cNext++;
+		} while (cNext < cTotal);
+
+		return data;
+	}
+
+	/**
+	 * store data into the id-th entry
+	 */
+	@Override
+	public int storeByteArray(final int id, final byte[] data) {
+		if (id == NewPage) {
+			Entry e = new Entry();
+			e.m_length = data.length;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;data&#39; (index: 1 | type: byte[). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;id&#39;. The same parameter type is validated here for example: njr_failed/url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/io/MemoryStorageManager.java:68. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).